refactor(store): migrate unit tests to Vitest

exequiel09 opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[x] Other... Please describe:

Updates to testing to ensure compatibility to `vitest`
```

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

The current behavior is that tests are running perfectly fine on `jest`. This PR ensures that it runs smoothly on `vitest` without any errors.

Closes #5018 

## What is the new behavior?

No new behavior. Just ensures to maintain parity when it runs using `vitest`

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

exequiel09 reviewed (commented):
Self-comment

exequiel09 commented on modules/store/spec/selector.spec.ts:
```diff
@@ -562,13 +559,13 @@ describe('Selectors', () => {
 
           expect(warnSpy).toHaveBeenCalled();
-          expect(warnSpy.calls.mostRecent().args[0]).toMatch(
+          expect(warnSpy.mock.calls[warnSpy.mock.calls.length - 1][0]).toMatch(
```
I think `lastCall` will be much cleaner https://vitest.dev/api/mock.html#mock-lastcall

exequiel09 commented on modules/store/migrations/8_0_0-rc/index.spec.ts:
```diff
@@ -231,6 +232,7 @@ describe('Migration to version 8.0.0 rc', () => {
           },
         }),
       },
+      // TODO this throws when migrated to vitest
```
Done in https://github.com/ngrx/platform/pull/5036/commits/a677a95cb2b22e21a0fb89355cdb837e62543ae8

exequiel09 commented:
Only type errors are left which I didn't touch and just appeared when migrated to `vitest`.

1. https://github.com/ngrx/platform/actions/runs/19849232805/job/56872554303?pr=5036#step:8:407
2. https://github.com/ngrx/platform/actions/runs/19849232805/job/56872554303?pr=5036#step:8:382

Thinking of how to fix this 🤔 

- option 1 = https://vitest.dev/config/dangerouslyignoreunhandlederrors.html 🤷🏻

exequiel09 commented:
Would love to ask for some feedback on the PR @markostanimirovic and @brandonroberts especially on the issues that arised upon migrating to `vitest`. Thank you.

exequiel09 commented on modules/store/spec/store.spec.ts:
```diff
@@ -277,8 +280,8 @@ describe('ngRx Store', () => {
+      vi.spyOn(dispatcher, 'next');
+      vi.spyOn(dispatcher, 'error');
 
       store.next(<any>1);
```
The `store.next` invocation can be wrapped in an anonymous fn and can be passed to the assertion with the expectation of throwing an error. However for the `store.error` it's  a different issue as you can't do that and it's treated by vitest as an uncaught exception.

timdeschryver reviewed (commented):
The remaining CI errors are not caused by this PR => they should be resolved in #5044.

From looking at the PR, I think you addressed all of the issues yourself, or is there still something left where you're waiting for input? 
The failing type tests can be fixed by updating the timeout to `8_000` within the describe block (this will be passed down to each test case), similar to what's done in #5044.

exequiel09 commented:
> The failing type tests can be fixed by updating the timeout to `8_000` within the describe block (this will be passed down to each test case), similar to what's done in #5044.

Addressed this in https://github.com/ngrx/platform/pull/5036/commits/35b9962c71e83374ffcb7139fab8f590f918a6de

exequiel09 commented:
> The remaining CI errors are not caused by this PR => they should be resolved in #5044.

There are 2 errors that's related to vitest uncaught exceptions.

1.  I mentioned in this comment as problem item 1 (https://github.com/ngrx/platform/pull/5036#issuecomment-3600383044)
<img width="2812" height="768" alt="Screenshot 2025-12-15 at 9 25 42 AM" src="https://github.com/user-attachments/assets/70a3fe39-2d48-43a2-a881-aee4e555545f" />

2. I mention in this comment as problem item 2 (https://github.com/ngrx/platform/pull/5036#issuecomment-3600383044)
<img width="2790" height="680" alt="Screenshot 2025-12-15 at 9 27 12 AM" src="https://github.com/user-attachments/assets/03777314-3a67-4c7b-960b-3bd145fc4a99" />


> From looking at the PR, I think you addressed all of the issues yourself, or is there still something left where you're waiting for input?

@timdeschryver  I was looking for inputs on how to resolve this. Also, I tried to propose solutions in the comment (https://github.com/ngrx/platform/pull/5036#issuecomment-3600383044), that needs to be discussed since the solutions I have thought of might not be up to the team's liking.

timdeschryver commented on modules/store/spec/store.spec.ts:
```diff
@@ -280,8 +280,8 @@ describe('ngRx Store', () => {
-      vi.spyOn(dispatcher, 'next');
-      vi.spyOn(dispatcher, 'error');
+      vi.spyOn(dispatcher, 'next').mockImplementationOnce(() => void 0);
+      vi.spyOn(dispatcher, 'error').mockImplementationOnce(() => void 0);
```
Yea, likewise 😅

exequiel09 commented:
> We can publish the PR and I'll review it once more with a fresh head, I suppose it's good to be merged though.

Hey @timdeschryver I removed the "draft" status and changed the title to use the type `refactor` instead of `wip`.

exequiel09 commented:
> I just rebased the PR just moments ago based on the latest commits on the main branch. Did something change again or got merged?

Checked. PR is still on top of the commits of `main` branch.

<img width="1770" height="272" alt="Screenshot 2025-12-16 at 2 57 42 PM" src="https://github.com/user-attachments/assets/74336e0f-694b-49ec-bed1-857da8fa86f3" />

<img width="2732" height="540" alt="Screenshot 2025-12-16 at 2 58 54 PM" src="https://github.com/user-attachments/assets/6a4a354f-6964-4597-83ce-195b3bc744f8" />

exequiel09 commented:
I see one of the issues exceeded 8000ms. That's why something failed. 

https://github.com/ngrx/platform/actions/runs/20258235781/job/58164627027?pr=5036#step:8:1558
<img width="1414" height="120" alt="Screenshot 2025-12-16 at 3 00 16 PM" src="https://github.com/user-attachments/assets/ba8f52ed-5cb5-4c7f-a327-90bb6d4eaf15" />
